pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/PrimaryKey.java

```diff
@@ -37,7 +40,56 @@ public Object[] getValues() {
   }
 
   public byte[] asBytes() {
-    return SerializationUtils.serialize(_values);
+    int arraySize = 0;
+    byte[][] cache = new byte[_values.length][];
+    for (int i = 0; i < _values.length; i++) {
+      Object value = _values[i];
+      if (value instanceof Integer) {
```

Comment:
(minor) Consider reordering based on the how common it is used as primary key: INT, LONG, STRING, BYTES, FLOAT, DOUBLE, BIG_DECIMAL